src/ctap/apdu.rs

```diff
@@ -168,12 +168,17 @@ impl TryFrom<&[u8]> for APDU {
         }
         if payload.len() > 2 {
             // Lc is possibly three-bytes long
-            let extended_apdu_lc: usize = BigEndian::read_u16(&payload[1..]) as usize;
-            let extended_apdu_le_len: usize = if payload.len() > extended_apdu_lc {
-                payload.len() - extended_apdu_lc - 3
-            } else {
-                0
+            let extended_apdu_lc: usize = BigEndian::read_u16(&payload[1..3]) as usize;
+            if payload.len() < extended_apdu_lc + 3 {
+                return Err(ApduStatusCode::SW_WRONG_LENGTH);
+            }
+            let extended_apdu_le_len: usize = match payload.len() - extended_apdu_lc {
+                // There's some possible Le bytes at the end
+                2..=5 => payload.len() - extended_apdu_lc - 3,
```

Comment:
Is this intended? What if `payload.len() - extended_apdu_lc == 2`? Wouldn't you underflow?

Reply:
@kaczmarczyck the check a couple lines above would ensure that's not the case

Reply:
Ah, now the range is `0..=3` and it matches. The comparison seems to be a bit convoluted, still. What you basically want is:

`let extended_apdu_le_len = match payload.len() - extended_apdu_lc - 3`
`if not 0 <= extended_apdu_le_len <= 3 return WRONG_LENGTH`

But because of `usize`, you need an `if` and a `match`?

Reply:
@kaczmarczyck yeah the reason I was doing the `if` dance was to prevent subtraction overflows. Now I case them as unsigned integers and evaluate the expression only one. Let me know if this one is more readable.

Reply:
There is an
`if payload.len() < extended_apdu_lc + 3 {`
So why cast? Why not just

`let extended_apdu_le_len = payload.len() - extended_apdu_lc - 3;`
`if extended_apdu_le_len > 3 { return [...]`

Reply:
@kaczmarczyck `payload.len()` and `extended_apdu_lc` are `usize` so `payload.len() - extended_apdu_lc - 3` on first look feels like it might possible overflow. you are right though that in this case it's not possible for them to both be `0` because of the checks above. updated as per your suggested snippet

Reply:
If this makes you feel uneasy, you could also do something like
`let extended_apdu_le_len payload.len().checked_sub(extended_apdu_lc + 3).map_err(...)?;`
I'm happy either way.

Reply:
good idea, updated to use `checked_sub` and `unwrap_or` the result